modules/bar/volume/index.ts

```diff
@@ -2,48 +2,77 @@ import Gdk from 'gi://Gdk?version=3.0';
 const audio = await Service.import("audio");
 import { openMenu } from "../utils.js";
 import options from "options";
-import { Binding } from 'lib/utils.js';
 import { VolumeIcons } from 'lib/types/volume.js';
+import { Stream } from 'types/service/audio';
+import Separator from "types/widgets/separator";
+import Label from "types/widgets/label";
 
 const Volume = () => {
-    const icons: VolumeIcons = {
-        101: "󰕾",
+    const { label, input, output, hide_muted_label } = options.bar.volume;
+
+    const outputIcons: VolumeIcons = {
+        101: "󱄠",
         66: "󰕾",
         34: "󰖀",
         1: "󰕿",
         0: "󰝟",
     };
 
-    const getIcon = () => {
-        const icon: Binding<number> = Utils.merge(
-            [audio.speaker.bind("is_muted"), audio.speaker.bind("volume")],
-            (isMuted, vol) => {
-                return isMuted
-                    ? 0
-                    : [101, 66, 34, 1, 0].find((threshold) => threshold <= vol * 100) || 101;
-            },
-        );
+    const inputIcons: VolumeIcons = {
+        51: "󰍬",
+        1: "󰍮",
+        0: "󰍭",
+    };
 
-        return icon.as((i: number) => i !== undefined ? icons[i] : icons[101]);
+    const getIcon = (icons: VolumeIcons, volume: number, isMuted: (boolean | null)): string => {
+        const keys = Object.keys(icons).map(Number).reverse();
+        let icon: number;
+        if (isMuted !== false || Math.round(volume * 100) === 0) {
+            icon = 0;
+        } else {
+            icon = keys.find((threshold) => threshold <= Math.round(volume * 100)) ?? keys[0];
+        }
+        return icons[icon];
     };
 
-    const volIcn = Widget.Label({
-        hexpand: true,
-        label: getIcon(),
-        class_name: "bar-button-icon volume txt-icon bar",
-    });
+    const volIcn = (audio_type: Stream, icons: VolumeIcons, extra_class_name: string, showLabel: boolean, hideMutedLabel: boolean): Label<any> => {
+        const class_name = `bar-button-icon volume txt-icon bar ${extra_class_name}`;
+        return Widget.Label({
+            hexpand: true,
+            class_name
+        }).hook(audio_type, (self) => {
```

Comment:
```suggestion
        }).hook(audio_type, (self) => {
```
Can we move this hook and other similar ones out of here as well and into the main function? Then just consume it as a parameter here? So we don't nest bindings/hooks.